codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -267,21 +291,66 @@ protected void handlerRemoved0(ChannelHandlerContext ctx) throws Exception { }
         if (msg instanceof ByteBuf) {
             CodecOutputList out = CodecOutputList.newInstance();
+            ByteBuf data = (ByteBuf) msg;
+
+            byte state = STATE_DEFER_HANDLER_REMOVED | STATE_DEFER_INPUT_CLOSED;
             try {
-                ByteBuf data = (ByteBuf) msg;
-                first = cumulation == null;
-                if (first) {
-                    cumulation = data;
-                } else {
+
+                decodeState = state;
+                do {
+                    first = cumulation == null;
+                    if (first) {
+                        callDecode(ctx, data, out);
+                        if (decodeState != state || !data.isReadable()) {
+                            break;
+                        }
+                        cumulation = Unpooled.EMPTY_BUFFER;
+                    }
+
+                    // we permit our cumulators to be smart, and guarantee exactly one message's worth of data is
+                    // buffered before we switch back to using the input buffer, so that if we are waiting for a message
+                    // of known size we can allocate an accurately sized buffer, and read any remaining messages without
+                    // copying or composing
+                    int oldReadableBytes = cumulation.readableBytes();
+                    int newReadableBytes = data.readableBytes();
                     cumulation = cumulator.cumulate(ctx.alloc(), cumulation, data);
-                }
-                callDecode(ctx, cumulation, out);
-            } catch (DecoderException e) {
-                throw e;
+                    if (cumulation.readableBytes() == oldReadableBytes + newReadableBytes) {
+                        data = null;
+                        if (first) {
+                            break; // don't bother calling decode again if we've already passed it everything
+                        }
+                    }
+
+                    callDecode(ctx, cumulation, out);
+                    if (decodeState != state) {
+                        break;
+                    }
+
+                    if (!cumulation.isReadable()) {
+                        cumulation.release();
+                        cumulation = null;
+                    }
+                } while (data != null);
+
             } catch (Exception e) {
+                if (e instanceof DecoderException) {
+                    throw e;
+                }
                 throw new DecoderException(e);
             } finally {
-                if (cumulation != null && !cumulation.isReadable()) {
+                state = decodeState;
+                decodeState = STATE_INIT;
+
+                saveToCumulation(data);
+                if (hasState(state, STATE_HANDLER_REMOVED_PENDING)) {
+                    handlerRemoved(ctx);
+                }
+                if (hasState(state, STATE_INPUT_CLOSED_PENDING)) {
```

Comment:
shouldn't we handle this state first before `handlerRemoved` ?

Reply:
Perhaps, thanks for the insight - I wasn't too sure about this.  I'll make that change.  I also have an alternative idea for on approach that might on aggregate be sinmpler that I'll try out tomorrow, and we can choose between the two approaches.